src/adapters/surrealDB/index.ts

```diff
@@ -0,0 +1,189 @@
+	});
+
+	const filterExpr = query['$filter']
+		? `WHERE ${Object.entries(query['$filter'])
+				.map(([key, value]) => `${key} = ${query['$thing']}:\`${value}\``)
+				.join(',')}`
+		: '';
+
+	const result = `SELECT 
+    ${[...attributes.map(convertEntityId), ...entitiesQuery, ...relationsQuery].join(',')}
+  FROM ${thing} 
+  ${filterExpr}
+  FETCH ${entities.map((entity) => entity['$path']).join(',')} 
+  PARALLEL`;
+
+	return result;
+};
+
+const buildSurrealDbQuery: PipelineOperation<SurrealDbResponse> = async (req, res) => {
+	const { dbHandles, enrichedBqlQuery, schema } = req;
+
+	if (!enrichedBqlQuery) {
+		throw new Error('BQL request not parsed');
+	}
+	if (!dbHandles.surrealDB) {
+		throw new Error('missing SurrealDB in dbHandles');
+	}
+
+	const connector = req.config.dbConnectors.find((connector) => connector.provider === 'surrealDB');
+
+	if (!connector) {
+		throw new Error('missing SurrealDB config');
+	}
+
+	const mapItem = dbHandles.surrealDB.get(connector.id);
+
+	if (!mapItem) {
+		throw new Error(`missing SurrealDB client with id of ${connector.id}`);
+	}
+
+	const { client } = mapItem;
+
+	const results = await Promise.all(
+		(req.enrichedBqlQuery as Array<EnrichedBqlQuery>).map(async (query, idx) => {
+			let queryResult: Array<Record<string, unknown>> | undefined;
+
+			if (query['$thingType'] !== 'entity') {
+				// NOTE relations from testSchema contains hyphen at the moment. Remove case transformation, once we use camelcase for all tables
+				const thing = pascal(query.$thing);
+
+				const generatedQuery = buildQuery(thing, query);
+
+				const queryRes: Array<
+					Record<string, unknown> & {
+						id: string;
+					}
+				> = await client.query(generatedQuery);
+
+				queryResult = queryRes.map((item) => ({
+					$id: item.id,
+					$thing: trainCase(thing),
+					$thingType: 'relation',
+					...item,
+					id: item.id,
+					[QueryPath]: enrichedBqlQuery[0][QueryPath],
+				}));
+			} else {
+				const subtypes = [
+					query['$thing'],
+					...getSubtype(req.schema, query['$thingType'] === 'entity' ? 'entities' : 'relations', query['$thing']),
+				];
+
+				const payload = await Promise.all(
+					subtypes.map(async (subtype) => {
+						const queryRes: Array<
+							Record<string, unknown> & {
+								id: string;
+							}
+						> = await client.query(buildQuery(subtype, query));
+
+						return queryRes.map((item) => ({
+							$id: dash(item.id),
+							$thing: subtype,
+							$thingType: 'entity',
+							...item,
+							id: dash(item.id),
+							[QueryPath]: enrichedBqlQuery[0][QueryPath],
+						}));
+					}),
+				);
+
+				queryResult = payload.flat();
+			}
+
+			if (!queryResult) {
+				throw new Error('empty query result');
+			}
+
+			return queryResult.map(handleCardinality(schema, query));
+		}),
+	);
+
+	if (req.enrichedBqlQuery.length > 1) {
+		throw new Error('batch query unimplemented');
```

Comment:
Do you think it will be easy to implement batched queries in surrealDB? Would that work in a single transaction?

Reply:
Yes I think that is possible, just that it is not implemented yet.